drivers/gpio/gpio_pcal6416a.c

```diff
@@ -0,0 +1,269 @@
+{
+	int rc;
+
+	LOG_DBG("writing to register 0x%02X value 0x%02X", register_address, value);
+	rc = i2c_reg_write_byte_dt(i2c, register_address, value);
+
+	if (rc != 0) {
+		LOG_ERR("unable to write to register 0x%02X, error %i", register_address, rc);
+	}
+
+	return rc;
+}
+
+static int pcal6416a_i2c_read(const struct i2c_dt_spec *i2c, uint8_t register_address,
+			      uint8_t *value)
+{
+	int rc;
+
+	rc = i2c_reg_read_byte_dt(i2c, register_address, value);
+	LOG_DBG("reading from register 0x%02X value 0x%02X", register_address, *value);
+
+	if (rc != 0) {
+		LOG_ERR("unable to read from register 0x%02X, error %i", register_address, rc);
+	}
+
+	return rc;
+}
+
+static int pcal6416a_pins_cfg_apply(const struct i2c_dt_spec *i2c,
+				    const struct pcal64xxa_pins_cfg *pins_cfg)
+{
+	int rc;
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_SELECT_0,
+				 (uint8_t)pins_cfg->pull_ups_selected);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_SELECT_1,
+				 (uint8_t)(pins_cfg->pull_ups_selected >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_ENABLE_0,
+				 (uint8_t)pins_cfg->pulls_enabled);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_PULL_UP_DOWN_ENABLE_1,
+				 (uint8_t)(pins_cfg->pulls_enabled >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_OUTPUT_PORT_0, (uint8_t)pins_cfg->outputs_high);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_OUTPUT_PORT_1,
+				 (uint8_t)(pins_cfg->outputs_high >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_CONFIGURATION_0,
+				 (uint8_t)pins_cfg->configured_as_inputs);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_write(i2c, PCAL6416A_REG_CONFIGURATION_1,
+				 (uint8_t)(pins_cfg->configured_as_inputs >> 8));
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int pcal6416a_inputs_read(const struct i2c_dt_spec *i2c, uint16_t *int_sources,
+				 uint16_t *input_port)
+{
+	int rc;
+	uint8_t value_low;
+	uint8_t value_high;
+
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INTERRUPT_STATUS_0, &value_low);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INTERRUPT_STATUS_1, &value_high);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	*int_sources = value_low | (value_high << 8);
+
+	/* This read also clears the generated interrupt if any. */
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INPUT_PORT_0, &value_low);
+	if (rc != 0) {
+		return -EIO;
+	}
+
+	rc = pcal6416a_i2c_read(i2c, PCAL6416A_REG_INPUT_PORT_1, &value_high);
+	if (rc != 0) {
+		LOG_ERR("failed to read input port: %d", rc);
+		return -EIO;
+	}
+
+	*input_port = value_low | (value_high << 8);
+
+	return 0;
+}
+
+static int pcal6416a_outputs_write(const struct i2c_dt_spec *i2c, uint16_t outputs)
```

Comment:
Please add the complementary `pcal6408a_outputs_write`

Reply:
done